src/recap.c

```diff
@@ -0,0 +1,490 @@
+    log_trace("\tpcap_file_header version_minor = %d",
+              pctx->pcap_header.version_minor);
+    log_trace("\tpcap_file_header snaplen = %d", pctx->pcap_header.snaplen);
+    log_trace("\tpcap_file_header linktype = %d", pctx->pcap_header.linktype);
+    pctx->data_size = 0;
+    pctx->state = PCAP_STATE_READ_PKT_HEADER;
+  } else if (read_size == 0) {
+    log_trace("No data received");
+    pctx->state = PCAP_STATE_FIN;
+  }
+
+  return 0;
+}
+
+int process_pkt_header_state(struct pcap_stream_context *pctx) {
+  ssize_t read_size = 0;
+  ssize_t pkt_header_size = (ssize_t)sizeof(struct pcap_pkthdr32);
+
+  size_t len = (pkt_header_size > pctx->data_size)
+                   ? pkt_header_size - pctx->data_size
+                   : 0;
+
+  if ((read_size = read_pcap(pctx, len)) < 0) {
+    log_error("read_pcap fail");
+    return -1;
+  }
+
+  if (pctx->data_size >= pkt_header_size) {
+    log_trace("Received pkt header:");
+    os_memcpy(&pctx->pkt_header, pctx->pcap_data, pkt_header_size);
+    log_trace("\tpcap_pkthdr ts_sec = %llu", pctx->pkt_header.ts_sec);
+    log_trace("\tpcap_pkthdr ts_usec = %llu", pctx->pkt_header.ts_usec);
+    log_trace("\tpcap_pkthdr caplen = %llu", pctx->pkt_header.caplen);
+    log_trace("\tpcap_pkthdr len = %llu", pctx->pkt_header.len);
+
+    if (pctx->pkt_header.caplen > pctx->pkt_header.len) {
+      log_error("caplen > len");
+      return -1;
+    }
+
+    pctx->data_size = 0;
+    pctx->state = PCAP_STATE_READ_PACKET;
+  } else if (read_size == 0) {
+    log_trace("No data received");
+    pctx->state = PCAP_STATE_FIN;
+  }
+
+  return 0;
+}
+
+void get_packet_header(struct pcap_stream_context *pctx,
+                       struct pcap_pkthdr *header) {
+  header->ts.tv_sec = pctx->pkt_header.ts_sec;
+  header->ts.tv_usec = pctx->pkt_header.ts_usec;
+  header->caplen = pctx->pkt_header.caplen;
+  header->len = pctx->pkt_header.len;
+}
+
+void free_packets(UT_array *packets) {
+  struct tuple_packet *p = NULL;
+  while ((p = (struct tuple_packet *)utarray_next(packets, p)) != NULL) {
+    free_packet_tuple(p);
+  }
+}
+
+int save_sqlite_packet(sqlite3 *db, UT_array *packets) {
+  struct tuple_packet *p = NULL;
+  while ((p = (struct tuple_packet *)utarray_next(packets, p)) != NULL) {
+    if (save_packet_statement(db, p) < 0) {
+      log_error("save_packet_statement fail");
+      return -1;
+    }
+  }
+
+  return 0;
+}
+
+int save_packet(struct pcap_stream_context *pctx) {
+  const char *ltype = pcap_datalink_val_to_name(pctx->pcap_header.linktype);
+  struct pcap_pkthdr header;
+  uint8_t *packet = (uint8_t *)pctx->pcap_data;
+
+  char cap_id[MAX_RANDOM_UUID_LEN];
+  generate_radom_uuid(cap_id);
+  get_packet_header(pctx, &header);
+
+  int npackets;
+  UT_array *packets = NULL;
+  utarray_new(packets, &tp_list_icd);
+
+  if ((npackets = extract_packets(ltype, &header, packet, pctx->ifname, cap_id,
+                                  packets)) < 0) {
```

Comment:
Defining vars and initiliasing/using them in the same statement.
 
```suggestion
  const char *ltype = pcap_datalink_val_to_name(pctx->pcap_header.linktype);
  uint8_t *packet = (uint8_t *)pctx->pcap_data;

  char cap_id[MAX_RANDOM_UUID_LEN];
  generate_radom_uuid(cap_id);
  
  struct pcap_pkthdr header;
  get_packet_header(pctx, &header);

  UT_array *packets = NULL;
  utarray_new(packets, &tp_list_icd);

  int npackets = extract_packets(ltype, &header, packet, pctx->ifname, cap_id,
                                 packets);  
  if (npackets < 0) {
```